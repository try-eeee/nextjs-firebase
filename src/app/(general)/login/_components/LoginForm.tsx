"use client";

import { FormEvent, useState } from "react";
import Link from "next/link";
import { useRouter } from "next/navigation";
import { getAuth, signInWithEmailAndPassword } from "firebase/auth";
import { getFirebaseAuth } from "@/firebaseClient";
import { Input } from "@/components/input/Input";
import { Button } from "@/components/button/Button";
import { Label } from "@/components/label/Label";
import styles from "./LoginForm.module.scss";

/**
 * ログインフォーム
 */
export default function LoginForm() {
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [error, setError] = useState("");
  const router = useRouter();

  async function handleSubmit(event: FormEvent) {
    event.preventDefault();
    setError("");
    try {
      const { clientApp } = getFirebaseAuth();
      if (!clientApp) {
        throw new Error("Missing app!");
      }
      const credential = await signInWithEmailAndPassword(
        getAuth(clientApp),
        email,
        password
      );
      const idToken = await credential.user.getIdToken();
      await fetch("/api/login", {
        headers: { Authorization: `Bearer ${idToken}` },
      });
      router.push("/");
    } catch (e) {
      setError("ログイン情報が正しくありません。");
    }
  }

  return (
    <main>
      <h1 className={styles.title}>ログイン</h1>
      <form onSubmit={handleSubmit} className={styles.form} action="#">
        <div>
          <Label htmlFor="email" align="left">
            メールアドレス
          </Label>

          <Input
            type="email"
            name="email"
            value={email}
            onChange={(e) => setEmail(e.target.value)}
            id="email"
            className=""
            placeholder="name@company.com"
            required
          />
        </div>
        <div className={styles.mb5rem}>
          <Label htmlFor="password" align="left">
            パスワード
          </Label>
          <Input
            type="password"
            name="password"
            value={password}
            onChange={(e) => setPassword(e.target.value)}
            id="password"
            placeholder="••••••••"
            className=""
            required
          />
        </div>
        {error && (
          <div className={styles.error} role="alert">
            <span className="">{error}</span>
          </div>
        )}
        <Button type="submit" className={styles.loginButton}>
          ログイン
        </Button>
        <p>
          まだ会員登録がお済みでないですか?
          <br />
          <Link href="/register" className="">
            新規会員登録は こちら
          </Link>
        </p>
      </form>
    </main>
  );
}
