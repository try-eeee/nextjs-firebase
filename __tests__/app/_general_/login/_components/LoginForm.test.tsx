import React from "react";
import { render } from "@testing-library/react";
import LoginForm from "@/app/(general)/login/_components/LoginForm";

jest.mock("next/navigation", () => ({
  useRouter: () => {
    push: jest.fn();
  },
}));

describe("LoginForm UnitTest", () => {
  test("Snapshot Test Of LoginForm", () => {
    const { asFragment } = render(<LoginForm />);
    expect(asFragment()).toMatchSnapshot();
  });
});
