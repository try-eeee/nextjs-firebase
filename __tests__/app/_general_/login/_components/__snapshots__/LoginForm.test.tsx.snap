// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm UnitTest Snapshot Test Of LoginForm 1`] = `
<DocumentFragment>
  <main>
    <h1
      class="title"
    >
      ログイン
    </h1>
    <form
      action="#"
      class="form"
    >
      <div>
        <label
          class="label left"
          for="email"
        >
          メールアドレス
        </label>
        <input
          class="input"
          id="email"
          name="email"
          placeholder="name@company.com"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="mb5rem"
      >
        <label
          class="label left"
          for="password"
        >
          パスワード
        </label>
        <input
          class="input"
          id="password"
          name="password"
          placeholder="••••••••"
          required=""
          type="password"
          value=""
        />
      </div>
      <button
        class="button default size-S"
        type="submit"
      >
        ログイン
      </button>
      <p>
        まだ会員登録がお済みでないですか?
        <br />
        <a
          class="registerLink"
          href="/register"
        >
          新規会員登録は 
          <span>
            こちら
          </span>
        </a>
      </p>
    </form>
  </main>
</DocumentFragment>
`;
